### PR TITLE
Apex Trigger

### PR DESCRIPTION
**Pull request by rathoreprem**
1. Not Allow any teacher to insert/update if that teacher is teaching Hindi (Use subject 
    multi select picklist).
2. Not allow any class to delete if there are more than one Female students.
3. Not allow insert student if class already reached MaxLimit.(Without using MyConnt 
    field).
4. When insert/update any student’s in class then update MyCount Accordingly(Use 
    MyCount as a number field).
5. In Opportunity, If the stage is changed from another value to CLOSED_WON or 
    CLOSED_LOST,populates the Close Date field with Today().
6. Create a new PickList “Custom Status” in Class object.(New,Open,Close,Reset) values. 
    When this field changed and value is “Reset” now then delete all associated students 
    with related Class.

### CODE REVIEW

File: Apex_Trigger/ClassDeletion.trigger [hunk 1]
```diff
@@ -0,0 +1,12 @@
+trigger ClassDeletion on Class__c (before delete) 
+{
+		AggregateResult[] r = [SELECT Class__r.id classId, COUNT(id) NumberOfGirls FROM Student__C WHERE Sex__c = 'Female' AND Class__c IN:trigger.old GROUP BY Class__r.id];
```
Comment (meta-vijay-khatri) [inline Apex_Trigger/ClassDeletion.trigger:3]: variable naming need to be improved

File: Apex_Trigger/ClassDeletion.trigger [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    for(AggregateResult re : r )
+    {
+         if(integer.valueof(re.get('NumberOfGirls')) > 1)
+         {
+            ID clsId = re.get('classId').toString();
+            Trigger.oldMap.get(clsId).addError('Cannot delete Class because in this class more than 1 girls exist.');
```
Comment (meta-vijay-khatri) [inline Apex_Trigger/ClassDeletion.trigger:9]: static final variables should hold messages

File: Apex_Trigger/MaxLimitTrigger.trigger [hunk 3]
```diff
@@ -0,0 +1,19 @@
+trigger MaxLimitTrigger on Student__c (before insert) 
+{
+	if(Trigger.isInsert) 
+    {
+       List<ID> classIds = new List<ID>();
+       for(Student__c std : Trigger.New) 
+       {
+           classIds.add(std.Class__c);
+       }
+       List<Class__c> ClassList = [Select NumberOfStudents__c,MaxSize__c,(Select Id from Students__r) from Class__c where Id IN :classIds];
```
Comment (meta-vijay-khatri) [inline Apex_Trigger/MaxLimitTrigger.trigger:10]: variable name should start with lower case

File: Apex_Trigger/MaxLimitTrigger.trigger [hunk 4]
```diff
@@ -0,0 +1,19 @@
+trigger MaxLimitTrigger on Student__c (before insert) 
+{
+	if(Trigger.isInsert) 
+    {
+       List<ID> classIds = new List<ID>();
```
Comment (meta-vijay-khatri) [inline Apex_Trigger/MaxLimitTrigger.trigger:5]: it's better to use Set<Id> here

File: Apex_Trigger/MaxLimitTrigger.trigger [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+       for(Student__c std : Trigger.New) 
+       {
+           classIds.add(std.Class__c);
+       }
+       List<Class__c> ClassList = [Select NumberOfStudents__c,MaxSize__c,(Select Id from Students__r) from Class__c where Id IN :classIds];
+       for(Class__c cls : ClassList) 
+       {
+           if(cls.MaxSize__c == cls.NumberOfStudents__c) 
+           {
+               Trigger.New[0].addError('Class size is Full. Please select another class');
```
Comment (meta-vijay-khatri) [inline Apex_Trigger/MaxLimitTrigger.trigger:15]: this will only add exception to first student.

File: Apex_Trigger/MaxLimitTrigger.trigger (hunk 3 above)
Comment (meta-vijay-khatri) [inline Apex_Trigger/MaxLimitTrigger.trigger:10]: where are we using `(Select Id from Students__r)`

File: Apex_Trigger/StudentCountTrigger.trigger [hunk 6]
```diff
@@ -0,0 +1,10 @@
+trigger StudentCountTrigger on Student__c (after insert,after update) 
+{
+	List<class__c> listOfClassesToUpdate=new List<class__c>();
```
Comment (meta-vijay-khatri) [inline Apex_Trigger/StudentCountTrigger.trigger:3]: space around ` = `  (operators)

File: Apex_Trigger/StudentCountTrigger.trigger [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+    listOfClassesToUpdate=[select id,mycount__c,(select id from students__r) from class__c where id in (select class__c from student__c where id in:Trigger.New)];
+    for(class__c classToUpdate:listOfClassesToUpdate)  
+    {
+        classToUpdate.mycount__c=classToUpdate.students__r.size();
```
Comment (meta-vijay-khatri) [inline Apex_Trigger/StudentCountTrigger.trigger:7]: have you verified this?

File: Apex_Trigger/StageOfOpportunity.trigger [hunk 8]
```diff
@@ -0,0 +1,14 @@
+trigger StageOfOpportunity on Opportunity (before update) 
+{
+    for(Opportunity opp : Trigger.new)
+    {
+        if(opp.StageName == 'CLOSED WON' || opp.StageName == 'CLOSED LOST')
+        {
+            Opportunity oldOpp = trigger.oldMap.get(opp.Id);
+            if(!(oldOpp.StageName == 'CLOSED WON' || oldOpp.StageName == 'CLOSED LOST'))
```
Comment (meta-vijay-khatri) [inline Apex_Trigger/StageOfOpportunity.trigger:8]: can we combine these two conditions ?

File: Apex_Trigger/CustomStatus.trigger [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+                cls.MyCount__c = 0;
+            }
+        }
+      }
+       else
+       {  
+            Set<Id> classSet = new Set<Id>();
+            for(Class__c cls : Trigger.new)
+            {
+                if(cls.Custom_Status__c == 'Reset')
```
Comment (meta-vijay-khatri) [inline Apex_Trigger/CustomStatus.trigger:18]: it's just checking if the value is `Reset` it's not checking if it's changed

**Review by meta-vijay-khatri — CHANGES_REQUESTED** [review]
we will change this to trigger handler pattern

File: Apex_Trigger/ClassDeletion.trigger (hunk 2 above)
Comment (rathoreprem) [reply Apex_Trigger/ClassDeletion.trigger:9]: done

File: Apex_Trigger/MaxLimitTrigger.trigger (hunk 3 above)
Comment (rathoreprem) [reply Apex_Trigger/MaxLimitTrigger.trigger:10]: done

Comment (rathoreprem) [reply Apex_Trigger/MaxLimitTrigger.trigger:10]: remove it from query

File: Apex_Trigger/StudentCountTrigger.trigger (hunk 6 above)
Comment (rathoreprem) [reply Apex_Trigger/StudentCountTrigger.trigger:3]: done

File: Apex_Trigger/ClassDeletion.trigger (hunk 1 above)
Comment (rathoreprem) [reply Apex_Trigger/ClassDeletion.trigger:3]: done

File: Apex_Trigger/MaxLimitTrigger.trigger (hunk 4 above)
Comment (rathoreprem) [reply Apex_Trigger/MaxLimitTrigger.trigger:5]: done it.

File: Apex_Trigger/CustomStatus.trigger (hunk 9 above)
Comment (rathoreprem) [reply Apex_Trigger/CustomStatus.trigger:18]: Done it.

**Pull request merged**